extensions/http-curl/tests/C2MetricsTest.cpp

```diff
@@ -0,0 +1,209 @@
+    : HeartbeatHandler(std::move(configuration)),
+      metrics_updated_successfully_(metrics_updated_successfully),
+      replacement_config_(getReplacementConfigAsJsonValue(replacement_config_path)) {
+  }
+
+  void handleHeartbeat(const rapidjson::Document& root, struct mg_connection* conn) override {
+    switch (test_state_) {
+      case TestState::VERIFY_INITIAL_METRICS: {
+        verifyMetrics(root);
+        sendEmptyHeartbeatResponse(conn);
+        break;
+      }
+      case TestState::SEND_NEW_CONFIG: {
+        sendHeartbeatResponse("UPDATE", "configuration", "889348", conn, {{"configuration_data", replacement_config_}});
+        test_state_ = TestState::VERIFY_UPDATED_METRICS;
+        break;
+      }
+      case TestState::VERIFY_UPDATED_METRICS: {
+        verifyUpdatedMetrics(root);
+        sendEmptyHeartbeatResponse(conn);
+        break;
+      }
+    }
+  }
+
+ private:
+  enum class TestState {
+    VERIFY_INITIAL_METRICS,
+    SEND_NEW_CONFIG,
+    VERIFY_UPDATED_METRICS
+  };
+
+  static void sendEmptyHeartbeatResponse(struct mg_connection* conn) {
+    mg_printf(conn, "HTTP/1.1 200 OK\r\nContent-Type: text/plain\r\nContent-Length: 0\r\nConnection: close\r\n\r\n");
+  }
+
+  void verifyMetrics(const rapidjson::Document& root) {
+    auto initial_metrics_verified =
+      root.HasMember("metrics") &&
+      root["metrics"].HasMember("RuntimeMetrics") &&
+      root["metrics"].HasMember("LoadMetrics") &&
+      root["metrics"].HasMember("ProcessorMetrics");
+    if (initial_metrics_verified) {
+      test_state_ = TestState::SEND_NEW_CONFIG;
+    }
+  }
+
+  void verifyUpdatedMetrics(const rapidjson::Document& root) {
+    auto updated_metrics_verified =
+      root.HasMember("metrics") &&
+      root["metrics"].HasMember("RuntimeMetrics") &&
+      root["metrics"].HasMember("LoadMetrics") &&
+      !root["metrics"].HasMember("ProcessorMetrics") &&
+      verifyUpdatedRuntimeMetrics(root["metrics"]["RuntimeMetrics"]) &&
+      verifyUpdatedLoadMetrics(root["metrics"]["LoadMetrics"]);
+
+    if (updated_metrics_verified) {
+      metrics_updated_successfully_ = true;
+    }
+  }
+
+  static bool verifyRuntimeMetrics(const rapidjson::Value& runtime_metrics) {
+    return runtime_metrics.HasMember("deviceInfo") &&
+      runtime_metrics.HasMember("flowInfo") &&
+      runtime_metrics["flowInfo"].HasMember("versionedFlowSnapshotURI") &&
+      runtime_metrics["flowInfo"].HasMember("queues") &&
+      runtime_metrics["flowInfo"].HasMember("components") &&
+      runtime_metrics["flowInfo"]["queues"].HasMember("2438e3c8-015a-1000-79ca-83af40ec1997") &&
+      runtime_metrics["flowInfo"]["components"].HasMember("FlowController") &&
+      runtime_metrics["flowInfo"]["components"].HasMember("GetTCP") &&
+      runtime_metrics["flowInfo"]["components"].HasMember("LogAttribute");
+  }
```

Comment:
we have a `verifyJSON` test util, could that be used here? (maybe with some enhancements? e.g. "__any__" matching all values if we only care about field existence)

Reply:
I checked it, but I think in this case it's not really viable. There are a lot of environment or flow state specific metrics that cannot be checked for and exact expected json and I think it's better to check on these specific fields separately.